java/arcs/core/storage/Handle.kt

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.core.storage
+
+import arcs.core.crdt.CrdtData
+import arcs.core.crdt.CrdtOperation
+import arcs.core.crdt.internal.VersionMap
+
+/**
+ * The [Callbacks] interface is a simple stand-in for the callbacks that a [Handle] might want
+ * to use* to signal information back to the runtime.
+ *
+ * This is mostly a place-holder at the moment.
+ */
+interface Callbacks {
+    fun onUpdate(handle: Handle<*, *, *>)
+}
+
+/**
+ * Base implementation of Arcs handles on the JVM runtime.
+ *
+ * A handle is in charge of translating SDK data operations into the appropriate CRDT operation,
+ * and then forwarding them to a [StorageProxy].
+ *
+ * This handle implementation is different from the [Handle] hierarchy defined in [arcs.sdk]. This
+ * handle knows about CRDTs and how to interface with storage. Eventually, the SDK-side [Handle]
+ * instances will map to an instance based on this class in the runtime (similar to the current
+ * strategy for wasm, in which Kotlin-wasm particles map to an instance of a storageNG handle
+ * in the JS/TS runtime).
+ *
+ * Note: A possible eventual goal is that this becomes the single runtime [Handle] implementation,
+ * targetting any runtime platform that Kotlin targets. For the moment, though, this is just
+ * used by JVM hosts.
+ *
+ * Implementations using this base are assumed to be based on a [CrdtModel] type. Concrete
+ * subclasses should select a [CrdtData] and [CrdtOperation] type for the [Handle] when
+ * implementing a concrete version of it, and expose only the consumer type [T].
+ */
+abstract class Handle<Data : CrdtData, Op : CrdtOperation, T>(
+    val name: String,
+    val storageProxy: StorageProxy<Data, Op, T>
+) {
+
+    /** A list of current registered [Callbacks] instances */
+    private val callbacks = mutableListOf<Callbacks>()
+
+    /** Local copy of the [VersionMap] for the backing CRDT. */
+    var versionMap = VersionMap()
+        protected set
+
+    /** Add a new [Callback] to the list. */
+    fun addCallbacks(callbacks: Callbacks) {
+        this.callbacks.add(callbacks)
+    }
+
+    /** Notify all registered [Callbacks] instances that an update has occurred. */
+    protected fun notifyListeners() {
+        callbacks.forEach {
+            it.onUpdate(this)
+        }
+    }
+
+    /** Read value from the backing [StorageProxy], updating the internal clock copy. */
+    protected val value: T
+        get() = storageProxy.getParticleView().let { (value, versionMap) ->
+            this.versionMap = versionMap
+            value
+        }
```

Comment:
If it's just passing the value through in the `let`, consider using `also`, so  you don't need to re-return the value:

```kotlin
get() = 
    storageProxy.getParticleView().also { (value, versionMap) -> this.versionMap = versionMap }
```

Reply:
It's only returning the `value` field from the unpacked object. Something like this could work?
```
get() = storageProxy.getParticleView().also{ (_, versionMap) ->
    this.versionMap = versionMap
}.value
```